src/main/java/com/google/cloud/spanner/jdbc/JdbcSavepoint.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.jdbc;
+
+import java.sql.SQLException;
+import java.sql.Savepoint;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class JdbcSavepoint implements Savepoint {
+  private static final AtomicInteger COUNTER = new AtomicInteger();
+
+  static JdbcSavepoint named(String name) {
+    return new JdbcSavepoint(-1, name);
+  }
+
+  static JdbcSavepoint unnamed() {
+    int id = COUNTER.incrementAndGet();
+    return new JdbcSavepoint(id, String.format("s_%d", id));
+  }
+
+  private final int id;
+  private final String name;
+
+  private JdbcSavepoint(int id, String name) {
+    this.id = id;
+    this.name = name;
+  }
+
+  @Override
+  public int getSavepointId() throws SQLException {
+    JdbcPreconditions.checkState(this.id >= 0, "This is a named savepoint");
```

Comment:
Suggestion : It might be more readable to track named vs un-named through an explicit boolean variable say `isNamed` . Using the the ID's state to determine this seems like an over-use of ID's value beyond what its meant for. For example - If for any reason we change this ID's type from `integer` to `String`, it ideally shouldn't impact the logic to determine named vs un-named. But in this case we will need to modify the logic.

Reply:
I don't feel it's really necessary given the size of the class. Also; changing the type of ID from integer to anything else will not be possible, as the type is specified by the JDBC API (that is; `getSavepointId` must return an `int`).